clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -756,10 +756,15 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
             joinPrepareTimer.update();
         }
 
+        final SortedSet<TopicPartition> partitionsToRevoke = getPartitionsToRevoke(protocol, generation, memberId);
+
         // async commit offsets prior to rebalance if auto-commit enabled
         // and there is no in-flight offset commit request
-        if (autoCommitEnabled && autoCommitOffsetRequestFuture == null) {
-            autoCommitOffsetRequestFuture = maybeAutoCommitOffsetsAsync();
+        if (autoCommitEnabled) {
+            pausePartitions(partitionsToRevoke);
```

Comment:
It seems like this bug (and some of the complexity in this patch) is due to the fact that we do the auto-commit prior to revoking partitions. I wonder if that is really necessary. If we revoke first, then the partitions would be removed from `SubscriptionState` and we wouldn't have to worry about fetches for these partitions returning. Could that work as well?

Reply:
but if we don't commit these offsets, i wonder if that breaks the poll contract, i.e. we need to commit the acked data from the previous poll.  Note that, I believe some of the data could already be returned to the client from the previous poll loop.

Reply:
I think what Jason suggested, is change from "committing offset subscribed partitions -> revoking", to "revoking -> "committing offset subscribed partitions(including revoked partitions)". I don't think it is a good idea because that would open a door to allow consumer committing offsets for partitions it doesn't own.